src/parseTools.js

```diff
@@ -1335,8 +1335,60 @@ function from64(x) {
 }
 
 function to64(x) {
-  if (!MEMORY64) return x;
-  return `BigInt(${x})`;
+  if (MEMORY64) return /^\d+$/.test(x) ? `${x}n` : `BigInt(${x})`;
+  return x;
+}
+
+// Generates an expression that tests whether a pointer (either a BigInt or a signed int32 JS Number) is aligned to the given byte multiple.
+function isPtrAligned(ptr, alignment) {
+  if (MEMORY64) return /^\d+$/.test(alignment) ? `${ptr} % ${alignment}n == 0` : `${ptr} % BigInt(${alignment}) == 0`;
+  return `${ptr} % ${alignment} == 0`;
+}
+
+const MAX_MEMORY53 = 2 ** 53; // == 9007199254740992
+
+// Converts a pointer (either a BigInt or a signed int32 JS Number) in-place to an index on the heap (a BigInt or an unsigned JS Number).
+// N.B. in ASSERTIONS mode may generate two statements (in form "a;b;").
+function convertPtrToIdx(ptr, accessWidth) {
+  const assertPointerAlignment = ASSERTIONS ? `${isPtrAligned(ptr, accessWidth)};` : '';
+
+  let conversion;
+  if (MEMORY64) {
+    if (MAXIMUM_MEMORY > MAX_MEMORY53) {
+      conversion = /^\d+$/.test(accessWidth) ? `${ptr} >>= ${accessWidth}n` : `${ptr} >>= BigInt(${accessWidth})`;
+    } else {
+      conversion = `${ptr} = Number(${ptr}) / ${accessWidth}`;
+    }
+  } else if (MAXIMUM_MEMORY > 2 * 1024 * 1024 * 1024) {
+    conversion = `${ptr} >>>= ${accessWidth}`;
+  } else {
+    conversion = accessWidth ? `${ptr} >>= ${accessWidth}` : '';
+  }
+
+  return assertPointerAlignment + conversion;
+}
+
+// Returns a pointer (either a BigInt or a signed int32 JS Number) shifted to an index on the heap (a BigInt or an unsigned JS Number).
+function ptrToIdx(ptr, accessWidth) {
+  if (MEMORY64) {
+    if (MAXIMUM_MEMORY > MAX_MEMORY53) return /^\d+$/.test(accessWidth) ? `${ptr} >> ${accessWidth}n` : `${ptr} >> BigInt(${accessWidth})`;
```

Comment:
How about a helper function `isIntLiteral(x)` .. assuming that is what this regex is trying to achieve.  Its wan't obvious to me on first readying and its a repeated patter withing this file.